File: services/ingestion/src/logger.ts
// Structured JSON logger — never logs raw transcripts or PII.
// All log lines are newline-delimited JSON for ingestion by log aggregators.

type LogLevel = 'info' | 'warn' | 'error';

interface LogEntry {
  level: LogLevel;
  service: string;
  message: string;
  timestamp: string;
  [key: string]: unknown;
}

function log(level: LogLevel, message: string, context: Record<string, unknown> = {}): void {
  const entry: LogEntry = {
    level,
    service: 'eso-ingestion',
    message,
    timestamp: new Date().toISOString(),
    ...context,
  };
  process.stdout.write(JSON.stringify(entry) + '\n');
}

export const logger = {
  info: (message: string, context?: Record<string, unknown>) => log('info', message, context),
  warn: (message: string, context?: Record<string, unknown>) => log('warn', message, context),
  error: (message: string, context?: Record<string, unknown>) => log('error', message, context),
};

File: services/shared/logger.py
"""Structured JSON logger. Never log raw transcripts or PII — only call_id + metadata."""

from __future__ import annotations

import json
import sys
from datetime import datetime, timezone
from typing import Any


def _emit(level: str, service: str, message: str, **context: Any) -> None:
    entry: dict[str, Any] = {
        "level": level,
        "service": service,
        "message": message,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        **context,
    }
    sys.stdout.write(json.dumps(entry) + "\n")
    sys.stdout.flush()


class StructuredLogger:
    def __init__(self, service: str) -> None:
        self._service = service

    def info(self, message: str, **context: Any) -> None:
        _emit("info", self._service, message, **context)

    def warn(self, message: str, **context: Any) -> None:
        _emit("warn", self._service, message, **context)

    def error(self, message: str, **context: Any) -> None:
        _emit("error", self._service, message, **context)
